File: asm/src/lexer.rs
/*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/

use crate::{error::AsmError, token::Token};

pub struct Lexer {
    src: Vec<char>,
    pos: usize,
    line: usize,
}

impl Lexer {
    pub fn new(src: &str) -> Self {
        Self {
            src: src.chars().collect(),
            pos: 0,
            line: 1,
        }
    }

    pub fn tokenize(mut self) -> Result<Vec<(Token, usize)>, AsmError> {
        let mut out = Vec::new();
        loop {
            self.skip_whitespace_inline();

            // check for end of file
            if self.pos >= self.src.len() {
                out.push((Token::Eof, self.line));
                break;
            }
            let ch = self.src[self.pos];
            match ch {
                // comments
                ';' | '#' => {
                    self.skip_line();
                }
                '/' => {
                    if self.peek_next() == Some('/') {
                        self.skip_line();
                    }
                }

                '\n' => {
                    out.push((Token::Newline, self.line));
                    self.line += 1;
                    self.pos += 1;
                }

                '\r' => {
                    self.pos += 1;
                } // windows moment

                ',' => {
                    out.push((Token::Comma, self.line));
                    self.pos += 1;
                }

                '"' => {
                    let (s, line) = self.read_string()?;
                    out.push((Token::Str(s), line));
                }

                '.' => {
                    self.pos += 1; //skip dot
                    let name = self.read_ident();
                    out.push((Token::Directive(name.to_ascii_lowercase()), self.line));
                }

                '0'..='9' | '-' => {
                    let (tok, line) = self.read_number()?;
                    out.push((tok, line));
                }

                'a'..='z' | 'A'..='Z' | '_' => {
                    let line = self.line;
                    let ident = self.read_ident();

                    // if next is colon this is a label
                    self.skip_whitespace_inline();
                    if self.pos < self.src.len() && self.src[self.pos] == ':' {
                        self.pos += 1;
                        out.push((Token::Ident(ident), line));
                        out.push((Token::Colon, line));
                    } else {
                        // register
                        if let Some(tok) = resolve_register(&ident) {
                            out.push((tok, line));
                        } else {
                            out.push((Token::Ident(ident), line));
                        }
                    }
                }

                other => {
                    return Err(AsmError::new(
                        self.line,
                        format!("unexpected character '{}'", other),
                    ));
                }
            }
        }
        Ok(out)
    }

    fn skip_whitespace_inline(&mut self) {
        while self.pos < self.src.len() && matches!(self.src[self.pos], ' ' | '\t') {
            self.pos += 1;
        }
    }

    fn skip_line(&mut self) {
        while self.pos < self.src.len() && self.src[self.pos] != '\n' {
            self.pos += 1;
        }
    }

    fn peek_next(&self) -> Option<char> {
        self.src.get(self.pos + 1).copied()
    }

    fn read_ident(&mut self) -> String {
        let start = self.pos;
        while self.pos < self.src.len()
            && (self.src[self.pos].is_alphanumeric()
                || self.src[self.pos] == '_'
                || self.src[self.pos] == '.')
        {
            self.pos += 1;
        }
        self.src[start..self.pos].iter().collect()
    }

    fn read_number(&mut self) -> Result<(Token, usize), AsmError> {
        let line = self.line;
        let negative = self.src[self.pos] == '-';
        if negative {
            self.pos += 1;
        }
        if self.pos >= self.src.len() || !self.src[self.pos].is_ascii_digit() {
            return Err(AsmError::new(line, "expected digit after '-'"));
        }

        if self.src[self.pos] == '0' && self.pos + 1 < self.src.len() {
            match self.src[self.pos + 1] {
                'x' | 'X' => {
                    self.pos += 2;
                    let s: String = self.src[self.pos..]
                        .iter()
                        .take_while(|c| c.is_ascii_hexdigit())
                        .collect();
                    self.pos += s.len();
                    let n = i64::from_str_radix(&s, 16)
                        .map_err(|_| AsmError::new(line, format!("invalid hex literal '{}'", s)))?;
                    return Ok((Token::Int(if negative { -n } else { n }), line));
                }
                'b' | 'B' => {
                    self.pos += 2;
                    let s: String = self.src[self.pos..]
                        .iter()
                        .take_while(|&&c| c == '0' || c == '1')
                        .collect();
                    self.pos += s.len();
                    let n = i64::from_str_radix(&s, 2).map_err(|_| {
                        AsmError::new(line, format!("invalid binary literal '{}'", s))
                    })?;
                    return Ok((Token::Int(if negative { -n } else { n }), line));
                }
                _ => {}
            }
        }

        // read integer part
        let int_part: String = self.src[self.pos..]
            .iter()
            .take_while(|c| c.is_ascii_digit())
            .collect();
        self.pos += int_part.len();

        // check for decimals
        if self.pos < self.src.len() && self.src[self.pos] == '.' {
            self.pos += 1; // consume .
            let frac_part: String = self.src[self.pos..]
                .iter()
                .take_while(|c| c.is_ascii_digit())
                .collect();
            self.pos += frac_part.len();

            // exponent
            let mut exp_part = String::new();
            if self.pos < self.src.len() && matches!(self.src[self.pos], 'e' | 'E') {
                exp_part.push('e');
                self.pos += 1;
                if self.pos < self.src.len() && matches!(self.src[self.pos], '+' | '-') {
                    exp_part.push(self.src[self.pos]);
                    self.pos += 1;
                }
                let digits: String = self.src[self.pos..]
                    .iter()
                    .take_while(|c| c.is_ascii_digit())
                    .collect();
                self.pos += digits.len();
                exp_part.push_str(&digits);
            }

            let s = format!("{}.{}{}", int_part, frac_part, exp_part);
            let f: f64 = s
                .parse()
                .map_err(|_| AsmError::new(line, format!("invalid float literal '{}'", s)))?;
            return Ok((Token::Float(if negative { -f } else { f }), line));
        }

        // int
        let n: i64 = int_part
            .parse()
            .map_err(|_| AsmError::new(line, format!("invalid integer '{}'", int_part)))?;
        Ok((Token::Int(if negative { -n } else { n }), line))
    }
    fn read_string(&mut self) -> Result<(String, usize), AsmError> {
        let line = self.line;
        self.pos += 1; // skip opening quote
        let mut s = String::new();
        loop {
            if self.pos >= self.src.len() {
                return Err(AsmError::new(line, "unterminated string literal"));
            }
            match self.src[self.pos] {
                '"' => {
                    self.pos += 1;
                    break;
                }
                '\\' => {
                    self.pos += 1;
                    let esc = match self.src.get(self.pos) {
                        Some('n') => '\n',
                        Some('t') => '\t',
                        Some('r') => '\r',
                        Some('0') => '\0',
                        Some('\\') => '\\',
                        Some('"') => '"',
                        other => {
                            return Err(AsmError::new(
                                line,
                                format!("unknown escape '\\{}'", other.unwrap_or(&'?')),
                            ));
                        }
                    };
                    s.push(esc);
                    self.pos += 1;
                }
                c => {
                    s.push(c);
                    self.pos += 1;
                }
            }
        }
        Ok((s, line))
    }
}
fn resolve_register(name: &str) -> Option<Token> {
    let lower = name.to_ascii_lowercase();

    // numbered
    if let Some(rest) = lower.strip_prefix('r') {
        if let Ok(n) = rest.parse::<u8>() {
            if n < 16 {
                return Some(Token::Reg(n));
            }
        }
    }
    if let Some(rest) = lower.strip_prefix('f') {
        if let Ok(n) = rest.parse::<u8>() {
            if n < 16 {
                return Some(Token::FReg(n));
            }
        }
    }

    // aliases
    let reg = match lower.as_str() {
        "zero" => 0,
        "ra" => 1,
        "sp" => 2,
        "fp" => 3,
        "rv" => 4,
        "a0" => 5,
        "a1" => 6,
        "a2" => 7,
        "a3" => 8,
        "a4" => 9,
        "t0" => 10,
        "t1" => 11,
        "t2" => 12,
        "t3" => 13,
        "s0" => 14,
        "s1" => 15,
        _ => return try_freg_alias(&lower),
    };
    Some(Token::Reg(reg))
}

fn try_freg_alias(lower: &str) -> Option<Token> {
    let n = match lower {
        "fa0" => 0,
        "fa1" => 1,
        "fa2" => 2,
        "fa3" => 3,
        "ft0" => 4,
        "ft1" => 5,
        "ft2" => 6,
        "ft3" => 7,
        "fs0" => 8,
        "fs1" => 9,
        "fs2" => 10,
        "fs3" => 11,
        _ => return None,
    };
    Some(Token::FReg(n))
}

File: asm/src/lexer.c
/*
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include "asm.h"
#include "isa.h"
#include "strcasecmp.h"
#include <ctype.h>
#include <stdio.h>
#include <string.h>

static int is_ident_start(char c) {
    return isalpha((unsigned char) c) || c == '_';
}

static int is_ident_cont(char c) {
    return isalnum((unsigned char) c) || c == '_' || c == '.';
}

static char *resolve_register(const char *name, tok_type *out_type, int64_t *out_val) {
    // numbered r
    if ((name[0] == 'R' || name[0] == 'r') && isdigit((unsigned char) name[1])) {
        int n = atoi(name + 1);
        if (n >= 0 && n < 16 && name[2 + (n >= 10)] == '\0') {
            *out_type = TOK_REG;
            *out_val = n;
            return (char *) name;
        }
    }
    // numbered f
    if ((name[0] == 'F' || name[0] == 'f') && isdigit((unsigned char) name[1])) {
        int n = atoi(name + 1);
        if (n >= 0 && n < 16) {
            *out_type = TOK_FREG;
            *out_val = n;
            return (char *) name;
        }
    }

    // non floating aliases
    static const struct {
        const char *name;
        int reg;
    } gpr[] = {
        {"zero", 0}, {"ra", 1},  {"sp", 2},  {"fp", 3},  {"rv", 4},  {"a0", 5},
        {"a1", 6},   {"a2", 7},  {"a3", 8},  {"a4", 9},  {"t0", 10}, {"t1", 11},
        {"t2", 12},  {"t3", 13}, {"s0", 14}, {"s1", 15},
    };
    for (size_t i = 0; i < sizeof(gpr) / sizeof(*gpr); i++) {
        if (strcasecmp(name, gpr[i].name) == 0) {
            *out_type = TOK_REG;
            *out_val = gpr[i].reg;
            return (char *) name;
        }
    }

    // floating aliases
    static const struct {
        const char *name;
        int reg;
    } fpr[] = {
        {"fa0", 0}, {"fa1", 1}, {"fa2", 2}, {"fa3", 3}, {"ft0", 4},  {"ft1", 5},
        {"ft2", 6}, {"ft3", 7}, {"fs0", 8}, {"fs1", 9}, {"fs2", 10}, {"fs3", 11},
    };
    for (size_t i = 0; i < sizeof(fpr) / sizeof(*fpr); i++) {
        if (strcasecmp(name, fpr[i].name) == 0) {
            *out_type = TOK_FREG;
            *out_val = fpr[i].reg;
            return (char *) name;
        }
    }

    return nullptr;
}

int lex(const char *src, token_vec *out, arena *a, asm_error *err) {
    const char *p = src; // current position
    int line = 1;        // current line number

    while (*p) {
        // whitespace
        while (*p == ' ' || *p == '\t' || *p == '\r') {
            p++;
        }
        if (!*p) {
            break;
        }

        // comment
        if (*p == ';') {
            // skip to eol
            while (*p && *p != '\n') {
                p++;
            }
        }

        // newline
        if (*p == '\n') {
            token t = {TOK_NEWLINE, line, {0}};
            vec_push(out, t);
            p++;
            continue;
        }

        if (*p == '\r') {
            p++;
            continue;
        }

        if (*p == ',') {
            token t = {TOK_COMMA, line, {0}};
            vec_push(out, t);
            p++;
            continue;
        }

        if (*p == '"') {
            p++; // skip opening quote
            strbuf sb;
            strbuf_init(&sb);
            while (*p && *p != '"') {
                if (*p == '\\') {
                    p++;
                    switch (*p) {
                        case 'n':
                            strbuf_appendc(&sb, '\n');
                            break;
                        case 't':
                            strbuf_appendc(&sb, '\t');
                            break;
                        case 'r':
                            strbuf_appendc(&sb, '\r');
                            break;
                        case '0':
                            strbuf_appendc(&sb, '\0');
                            break;
                        case '\\':
                            strbuf_appendc(&sb, '\\');
                            break;
                        case '"':
                            strbuf_appendc(&sb, '"');
                            break;
                        default:
                            asm_err(err, line, "unknown escape '\\%c'", *p);
                            strbuf_free(&sb);
                            return 0;
                    }
                } else {
                    strbuf_appendc(&sb, *p);
                }
                p++;
            }
            if (*p != '"') {
                asm_err(err, line, "unterminated string literal");
                strbuf_free(&sb);
                return 0;
            }
            p++;

            size_t len = sb.len;
            char *copy = arena_alloc(a, len + 1);
            memcpy(copy, strbuf_get(&sb), len + 1);
            strbuf_free(&sb);

            token t = {TOK_STR, line, {0}};
            t.str = copy;
            vec_push(out, t);
            continue;
        }

        // directive
        if (*p == '.') {
            p++;
            const char *start = p;
            while (is_ident_cont(*p)) {
                p++;
            }
            size_t len = (size_t) (p - start);
            if (len == 0) {
                asm_err(err, line, "expected directive name after '.'");
                return 0;
            }
            char *name = arena_alloc(a, len + 1);
            memcpy(name, start, len);
            name[len] = '\0';
            for (size_t i = 0; i < len; i++) {
                name[i] = (char) tolower((unsigned char) name[i]);
            }
            token t = {TOK_DIRECTIVE, line, {0}};
            t.str = name;
            vec_push(out, t);
            continue;
        }

        // number
        if (isdigit((unsigned char) *p) || (*p == '-' && isdigit((unsigned char) p[1]))) {
            int negative = *p == '-';
            if (negative) {
                p++;
            }

            // hex
            if (*p == '0' && (p[1] == 'x' || p[1] == 'X')) {
                p += 2;
                const char *start = p;
                while (isxdigit((unsigned char) *p)) {
                    p++;
                }
                char buf[32] = {0};
                size_t len = (size_t) (p - start);
                if (len >= sizeof(buf)) {
                    asm_err(err, line, "hex literal too long");
                    return 0;
                }
                memcpy(buf, start, len);
                int64_t val = (int64_t) strtoll(buf, nullptr, 16);
                token t = {TOK_INT, line, {0}};
                t.ival = negative ? -val : val;
                vec_push(out, t);
                continue;
            }

            // binary
            if (*p == '0' && (p[1] == 'b' || p[1] == 'B')) {
                p += 2;
                const char *start = p;
                while (*p == '0' || *p == '1') {
                    p++;
                }
                char buf[64] = {0};
                size_t len = (size_t) (p - start);
                if (len >= sizeof(buf)) {
                    asm_err(err, line, "binary literal too long");
                    return 0;
                }
                memcpy(buf, start, len);
                int64_t n = (int64_t) strtoll(buf, NULL, 2);
                token t = {TOK_INT, line, {0}};
                t.ival = negative ? -n : n;
                vec_push(out, t);
                continue;
            }

            // decimal
            const char *start = p;
            while (isdigit((unsigned char) *p)) {
                p++;
            }

            // float
            if (*p == '.' && isdigit((unsigned char) p[1])) {
                p++;
                while (isdigit((unsigned char) *p)) {
                    p++;
                }
                if (*p == 'e' || *p == 'E') {
                    p++;
                    if (*p == '+' || *p == '-') {
                        p++;
                    }
                    while (isdigit((unsigned char) *p)) {
                        p++;
                    }
                }
                size_t len = (size_t) (p - start);
                char buf[64] = {0};
                if (len >= sizeof(buf)) {
                    asm_err(err, line, "float literal too long");
                    return 0;
                }
                memcpy(buf, start, len);
                double f = strtod(buf, NULL);
                token t = {TOK_FLOAT, line, {0}};
                t.fval = negative ? -f : f;
                vec_push(out, t);
                continue;
            }

            // integer
            size_t len = (size_t) (p - start);
            char buf[32] = {0};
            if (len >= sizeof(buf)) {
                asm_err(err, line, "integer literal too long");
                return 0;
            }
            memcpy(buf, start, len);
            int64_t n = strtoll(buf, NULL, 10);
            token t = {TOK_INT, line, {0}};
            t.ival = negative ? -n : n;
            vec_push(out, t);
            continue;
        }

        // ident or register
        if (is_ident_start(*p)) {
            const char *start = p;
            while (is_ident_cont(*p)) {
                p++;
            }
            size_t len = (size_t) (p - start);
            char *name = arena_alloc(a, len + 1);
            memcpy(name, start, len);
            name[len] = '\0';

            const char *q = p;
            while (*q == ' ' || *q == '\t') {
                q++;
            }

            // label
            if (*q == ':') {
                token ti = {TOK_IDENT, line, {0}};
                ti.str = name;
                token tc = {TOK_COLON, line, {0}};
                vec_push(out, ti);
                vec_push(out, tc);
                p = q + 1;
                continue;
            }

            // check for directive
            tok_type rtype;
            int64_t rval;
            if (resolve_register(name, &rtype, &rval)) {
                token t = {rtype, line, {0}};
                t.ival = rval;
                vec_push(out, t);
            } else {
                token t = {TOK_IDENT, line, {0}};
                t.str = name;
                vec_push(out, t);
            }
            continue;
        }

        asm_err(err, line, "unexpected character '%c'", *p);
        return 0;
    }

    token eof = {TOK_EOF, line, {0}};
    vec_push(out, eof);
    return 1;
}